File: reverse_list.cpp
#include "listf.h"

Node *reverse_list(Node *head)
{
    Node    *previous;
    Node    *next_elem;

    previous = NULL;
    while (head != NULL)
    {
        next_elem = head->next;
        head->next = previous;
        previous = head;
        head = next_elem;
    }
    return (previous);
}

File: go_alorithms/tasks/reverse_list.go
package tasks

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev, tmp *ListNode

	for head.Next != nil {
		prev = head
		head = head.Next
		prev.Next = tmp
		tmp = prev
	}
	return head
}
